### ISSUE TITLE
Keyboard layout changed callback

### ISSUE THREAD

**Issue by ghuser404** [glfw/glfw#1201]
Imagine a game shows a floating tutorial hint where it says "Press W to go forward". I would like to be notified when user switches their keyboard layout to another language, to show W equivalent in that layout. Otherwise, I would have to query the key name over and over.

**Comment by elmindreda** [comment]
I was hoping no one would bring this up until 4.0 when events should be cheaper (in terms of API surface) to add, but you're right. This is needed.

**Comment by NeatSketch** [comment]
@ghuser404 
There are keyboards with two layouts printed on the keys, Latin and Cyrillic. The two alphabets share some common letters, but they usually don't share the same keys. For example, the Cyrillic letter 'A' on a Bulgarian keyboard is on the same place as the English 'D'. Keyboard hotkeys and shortcuts are usually shown in English and the user usually expects that. Imagine showing a hint 'Press A'. The user will probably think the app/game means the English 'A'. But if your app means 'A' in currently selected keyboard layout, the user will try to press the wrong key. I actually saw someone in this situation, and the user was upset because "the controls won't work".

**Comment by ghuser404** [comment]
I am well aware of such keyboards. But think French, for example. Their 'A' is not where English 'A' is.

**Comment by NeatSketch** [comment]
@ghuser404
I know. How is it different from my example?

**Comment by ghuser404** [comment]
Because French keyboard doesn't have English letters printed on the keys, like Bulgarian one does. So if you have a French keyboard with French layout set in OS (because why wouldn't you) and a game tells you to press 'A', you will be pressing 'Q' (because that's what it is in US layout) and be more frustrated.

**Comment by NeatSketch** [comment]
French use AZERTY for both English and French, so if you request the user to press English 'A' they will press the key that both looks to them as 'A' and recognized by the OS as 'A'.

**Comment by ghuser404** [comment]
To ask user with French layout to press 'A', you have to call `glfwGetKeyName(GLFW_KEY_Q, 0);`.

**Comment by NeatSketch** [comment]
Yes, but that has nothing to do with the user's currently selected input language. What matters is the physical keyboard layout. To determine it you should detect which layout is selected for English and display the English letter that corresponds to the needed key. That way you don't have the problem that I explained and also don't have the problem that you explained.

**Comment by ghuser404** [comment]
Please provide code.
AFAIK, there is no such API available.

**Comment by ghuser404** [comment]
Actually, even if there was such API, imagine following scenario:
You want to save WASD as moving keys. If your game was to query where on your keyboard 'A' is positioned, it would map it to 'Q'. Yes, the game would be saying "Press 'A' to move left", but your 'A' will not be where you want it to be. For French keyboard you actually want your game to say "Use ZQSD to move" - not "WASD".

**Comment by NeatSketch** [comment]
> For French keyboard you actually want your game to say "Use ZQSD to move" - not "WASD".
Exactly what I meant your game should do.
My point was that it should not depend on the selected input language.

**Comment by ghuser404** [comment]
There is no such API that you're speaking of :)
Try writing some code and you'll get what I mean.

**Comment by NeatSketch** [comment]
Yes there is, just not in GLFW unfortunately.
I don't know how it works in other operating systems, but in Windows it's as follows:
There can be one or more input languages, and each of them can have one or more keyboard layouts.
For example, on a French PC the configuration is usually like this:

* Input language: French
  * Layout: AZERTY
* Input language: English
  * Layout: AZERTY

Another example: on a Russian PC the configuration is usually like this:

* Input language: Russian
  * Layout: ЙЦУКЕН
* Input language: English
  * Layout: QWERTY

The user can set up more languages or even more layouts under the same languages if they need it.

An app can determine which input languages and layouts are configured by calling a WinAPI function `GetKeyboardLayoutList`.
It returns a list of identifiers that correspond to all the pairs of configured input languages & layouts.
The low word of each value contains a Language Identifier for the input language and the high word contains a device handle to the physical layout of the keyboard.
I'm pretty sure there are APIs for the same thing in other operating systems too.

What I'm proposing is to query this list to determine which layouts are currently configured for English (usually it's just one) and then use it to correctly map the keys to their English letter representations. That way a game can show 'WASD' on a US keyboard and 'ZQSD' on a French keyboard. Also it does not depend on the currently selected input language (because it doesn't make sense) and doesn't confuse users with Cyrillic keyboards.

As for the code, I may try to implement this, at least for Windows. In that case I'll make a PR.

**Comment by ghuser404** [comment]
You're talking non-sense.

`GetKeyboardLayoutList` returns currently configured virtual keyboard layouts. It has nothing to do with physically printed keys on your keyboard.

**Comment by NeatSketch** [comment]
I know it returns configured virtual keyboard layouts. I don't know why you assumed I talked about physically printed keys.

**Comment by ghuser404** [comment]
Nothing is confusing in the docs.

You proposed for a French keyboard (where keys are printed AZERTY) to show "Press ZQSD to move", but gave the solution that will not work. User with French keyboard may have only US layout configured on their OS, so you won't be able to show the "ZQSD" instead of "WASD".

If you want to continue this discussion, I suggest you create a topic on GLFW forums and post a link here.

**Comment by X-Ryl669** [comment]
I'm french, and I don't get it.
On my keyboard, the keys are written ZQSD for what US consumer would see as WASD on her keyboard.
Whatever the keyboard layout I select (english or french), I expect the software to say "Press Z to go forward". I don't care if, with US layout selected, the "Z" key produces a "W" (basically, I expect the software to react on scan code not key name). So, I don't see the point of such callback.

The only issue I see here is if the software is started with US layout selected, then the software will tell "press W to move forward" while, physically, the letter "Z" is shown on that key, but, honestly, I think that users doing this (switching KB layout) are tech savvy enough to sort this out by themselves. As the doc says, you should save the scan code in your shortcut config, not the key name, since the scan code never changes. If you do what the doc says, then even if the user switched the layout **after** having saved the shortcut, the software will still work correctly (as expected by the user)

**Comment by ghuser404** [comment]
So your game's interface will depend on the keyboard layout the game was started with? Good game... /s
On a more serious note, this is how some big games do it, that's why you need such callback.

**Comment by X-Ryl669** [comment]
Honestly, I understand the issue you are trying to describe, but the example you gave is just not convincing.

When I'm playing a game, I don't care about the layout. I care about what the game says, if it says press "A" for something, I'll press the key that labelled "A" on my keyboard. If I press Alt + Shift during the game, I don't expect the game to stop working. When I'm editing the shortcut, yes, I'll care about the layout, but since I'm not 100% dumb, I'll set the layout for my actual keyboard for this (it would be crazy not to do so). From then, whatever the currently selected layout, I expect my shortcut to work (and this can only work if the game dev used scancode any not key name for those).

Please notice that if the user set the US layout on her FR keyboard, and the game says "press W to go forward", it's terrible UI (since the key labelled "W" is not the key generating "W" anymore). And because GLFW is not in the user mind, nor can GLFW ask for the native keyboard layout (another dumbness we supports since IBM PC), there is nothing GLFW could do to avoid this. The only sane behaviour is to expect that the user will use her keyboard native layout when it uses the application (at least when she create & save the shortcuts). Any different behaviour is meaningless and confusing.

Anyway, your issue is more important for another use case, that is, when the user is using a different layout to implement a different behaviour. Let's say you are writing a edit box, then you likely don't want to call getKeyName for every keypress to make sure the key's shortcut match the scan code you've remembered. Having such a callback would be a real lifesaver since you know you can cache your previous mapping until the callback is fired again.

**Comment by ghuser404** [comment]
> When I'm playing a game, I don't care about the layout. I care about what the game says, if it says press "A" for something, I'll press the key that labelled "A" on my keyboard.

If this was possible to do correctly, this issue would never have been opened.

> Please notice that if the user set the US layout on her FR keyboard, and the game says "press W to go forward", it's terrible UI (since the key labelled "W" is not the key generating "W" anymore). And because GLFW is not in the user mind, nor can GLFW ask for the native keyboard layout (another dumbness we supports since IBM PC), there is nothing GLFW could do to avoid this.

Here you answer your own question.

> The only sane behaviour is to expect that the user will use her keyboard native layout when it uses the application (at least when she create & save the shortcuts). Any different behaviour is meaningless and confusing.

You cannot expect this from user. Do you actually think to yourself "OK, I'm starting a game, I better switch back to that native keyboard layout or else my game UI will be screwed"?.. No!

Actually, I'm not surprised to hear this from someone who's French. I don't think you have to switch between French and English layouts while playing games. Now, imagine someone who talks to people both in English and in Russian at the same time, they switch layouts all the time.

> Please notice that if the user set the US layout on her FR keyboard, and the game says "press W to go forward", it's terrible UI (since the key labelled "W" is not the key generating "W" anymore).

Again, let me repeat myself, this is how big companies do UI in their games. You may not be a huge gamer yourself to know that. It is known as best practices, you should use it.

Honestly, I would like to stop this discussion. Enough has been said and discussed, and you always can come back to the older comments.

**Comment by h3xl3r** [comment]
So this issue has been closed (because of the frustrating discussion) but I am still in dire need for this callback and now wondering if it will never be added.

Just to be clear, I am not asking when this will happen, I just would like to know if the actual maintainers recognize this is needed and have "accepted" the issue (as @elmindreda's first comment seems to indicate), or if this is a "won't fix".

Is there maybe another (open) issue referencing this?

If such a thing will never be added I'll have to start planning a patched solution for my own applications as this is very much needed for proper handling at runtime.

Thanks for your time!
